File: 2023/typescript/src/day03.ts
{
  const input = `467..114..
  ...*......
  ..35..633.
  ......#...
  617*......
  .....+.58.
  ..592.....
  ......755.
  ...$.*....
  .664.598..`;

  console.clear();

  const lines = input.split("\n").map((line) => line.trim());
  const correctNumbers: number[] = [];
  const symbols = "*#$&=+-%@/";

  const gears: Map<string, number[]> = new Map();

  lines.forEach((line, lineIndex) => {
    const regex = /(\d+)/g;
    const matches = line.matchAll(regex);

    for (let match of matches) {
      const num = match[0];
      const index = match.index || 0;

      let isNumberAdjacentToAnySymbol = false;

      const startX = Math.max(index - 1, 0);
      const endX = Math.min(index + num.length, line.length - 1);
      const startY = Math.max(lineIndex - 1, 0);
      const endY = Math.min(lineIndex + 1, lines.length - 1);

      for (let y = startY; y <= endY; y++) {
        for (let x = startX; x <= endX; x++) {
          const char = lines[y][x];

          if (symbols.includes(char)) {
            isNumberAdjacentToAnySymbol = true;

            if (char == "*") {
              gears.set(
                `${x},${y}`,
                (gears.get(`${x},${y}`) ?? []).concat(parseInt(num))
              );
            }
          }
        }
      }

      if (isNumberAdjacentToAnySymbol) {
        correctNumbers.push(parseInt(num));
      }
    }
  });

  console.log({
    part1: correctNumbers.reduce((a, b) => a + b),
    part2: Array.from(gears.values())
      .filter((parts) => parts.length > 1)
      .reduce((a, b) => a + b.reduce((c, d) => c * d), 0),
  });
}

File: 2023/kotlin/src/main/kotlin/day03.kt
import kotlin.math.max
import kotlin.math.min

const val dayThreeInput = """467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598.."""

fun main() {
    val lines = dayThreeInput.split("\n")
    val symbols = "*#\\$&=+-%@/"

    val engineParts = mutableListOf<Int>()
    val gears: MutableMap<String, MutableList<Int>> = mutableMapOf()

    lines.forEachIndexed { lineIndex, line ->
        """(\d+)""".toRegex().findAll(line).forEach { matchResult ->
            var isAdjacentToSymbol = false

            val startX = max(matchResult.range.first - 1, 0)
            val endX = min(matchResult.range.last + 1, line.length - 1)
            val startY = max(lineIndex - 1, 0)
            val endY = min(lineIndex + 1, lines.count() - 1)

            for (x in startX..endX) {
                for (y in startY..endY) {
                    val currentChar = lines[y][x]
                    if (symbols.contains(currentChar)) {
                        isAdjacentToSymbol = true
                    }

                    if (currentChar == '*') {
                        val newValue = gears["$x,$y"] ?: mutableListOf()
                        newValue.add(matchResult.value.toInt())
                        gears["$x,$y"] = newValue
                    }
                }
            }

            if (isAdjacentToSymbol) {
                engineParts.add(matchResult.value.toInt())
            }
        }
    }

    println("part1: ${engineParts.sum()}")
    println("part2: ${gears.filter { it.value.count() > 1 }.values.sumOf { subList -> subList.reduce { a, b -> a * b } }}")
}